src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,370 @@
-    return um;
+    fclose(file);
+    curl_easy_cleanup(curl);
+    curl_global_cleanup();
+    return fileName;
 }
 
 bool UpgradeManager::HasNewSW()
 {
-    /// Check website, verify if sig is valid && SHA-256 is new
-    /// TBD
+    LOG_MARKER();
+
+    string pubKeyFileName = DownloadFile(PUBLIC_KEY_FILE_NAME);
+
+    if (pubKeyFileName.empty())
+    {
+        LOG_GENERAL(WARNING, "Cannot download public key file!");
+        return false;
+    }
+
+    string versionName = DownloadFile(VERSION_FILE_NAME);
+
+    if (versionName.empty())
+    {
+        LOG_GENERAL(WARNING, "Cannot download version file!");
+        return false;
+    }
+
+    vector<PubKey> pubKeys;
+    {
+        fstream pubKeyFile(PUBLIC_KEY_FILE_NAME, ios::in);
+        string pubKey;
 
-    return false;
+        while (getline(pubKeyFile, pubKey))
+        {
+            pubKeys.emplace_back(DataConversion::HexStrToUint8Vec(pubKey), 0);
+        }
+    }
+
+    string shaStr, sigStr;
+    {
+        fstream versionFile(VERSION_FILE_NAME, ios::in);
+        int line_no = 0;
+
+        /// Read SHA-256 hash
+        while (line_no != 14 && getline(versionFile, shaStr))
+        {
+            ++line_no;
+        }
+
+        /// Read signature
+        while (line_no != 16 && getline(versionFile, sigStr))
+        {
+            ++line_no;
+        }
+    }
+
+    const vector<unsigned char> sha = DataConversion::HexStrToUint8Vec(shaStr);
+    const unsigned int len = sigStr.size() / pubKeys.size();
+    vector<Signature> mutliSig;
+
+    for (unsigned int i = 0; i < pubKeys.size(); ++i)
+    {
+        mutliSig.emplace_back(
+            DataConversion::HexStrToUint8Vec(sigStr.substr(i * len, len)), 0);
+    }
+
+    /// Multi-sig verification
+    for (unsigned int i = 0; i < pubKeys.size(); ++i)
+    {
+        if (!Schnorr::GetInstance().Verify(sha, mutliSig.at(i), pubKeys.at(i)))
+        {
+            LOG_GENERAL(WARNING, "Multisig verification failed!");
+            return false;
+        }
+    }
+
+    return m_latestSHA != sha;
 }
 
 bool UpgradeManager::DownloadSW()
 {
-    /// Download SW from website, then update current SHA-256 value & curSWInfo
-    /// TBD
+    LOG_MARKER();
+
+    string versionName = DownloadFile(VERSION_FILE_NAME);
+
+    if (versionName.empty())
+    {
+        LOG_GENERAL(WARNING, "Cannot download version file!");
+        return false;
+    }
+
+    m_packageFileName = DownloadFile(PACKAGE_FILE_EXTENSION);
+
+    if (m_packageFileName.empty())
+    {
+        LOG_GENERAL(WARNING, "Cannot download package (.deb) file!");
+        return false;
+    }
+
+    uint32_t major, minor, fix, commit;
+    uint64_t upgradeDS;
+    string sha;
+    {
+        fstream versionFile(VERSION_FILE_NAME, ios::in);
+        int line_no = 0;
+        string line;
+
+        /// Read major version
+        while (line_no != 2 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
 
+        major = stoul(line);
```

Comment:
`stoul` could throw exceptions when no conversion can be performed.

https://en.cppreference.com/w/cpp/string/basic_string/stoul